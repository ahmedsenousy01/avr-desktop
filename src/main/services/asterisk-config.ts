/**
 * Token rendering utilities for Asterisk configuration templates.
 *
 * Supported tokens:
 *  - {{EXTERNAL_IP}}
 *  - {{SIP_PORT}}
 *  - {{RTP_START}}
 *  - {{RTP_END}}
 *  - {{CODECS}} (comma-separated)
 *  - {{DTMF_MODE}}
 *
 * Escaping:
 *  - Use \{{ and \}} in templates to output literal braces ({{ or }}) without triggering replacement.
 */
import { existsSync, promises as fs } from "node:fs";
import path from "node:path";

import type { AsteriskConfig } from "@shared/types/asterisk";
import { validateAsteriskConfig as validateAsteriskConfigSchema } from "@shared/types/asterisk";

/** Left delimiter for tokens in templates */
export const TOKEN_OPEN = "{{" as const;
/** Right delimiter for tokens in templates */
export const TOKEN_CLOSE = "}}" as const;

export type AsteriskTokenName = "EXTERNAL_IP" | "SIP_PORT" | "RTP_START" | "RTP_END" | "CODECS" | "DTMF_MODE";

/** All supported token names (for validation and discovery) */
export const ASTERISK_TOKENS: readonly AsteriskTokenName[] = [
  "EXTERNAL_IP",
  "SIP_PORT",
  "RTP_START",
  "RTP_END",
  "CODECS",
  "DTMF_MODE",
];

/** Build the delimited token string, e.g., "{{EXTERNAL_IP}}" */
export function makeToken(name: AsteriskTokenName): string {
  return `${TOKEN_OPEN}${name}${TOKEN_CLOSE}`;
}

function formatCodecs(codecs: AsteriskConfig["codecs"]): string {
  return codecs.join(",");
}

/** Build a map from delimited tokens to concrete string values for a given config. */
export function buildTokenMap(config: AsteriskConfig): Record<string, string> {
  return {
    [makeToken("EXTERNAL_IP")]: config.externalIp,
    [makeToken("SIP_PORT")]: String(config.sipPort),
    [makeToken("RTP_START")]: String(config.rtpStart),
    [makeToken("RTP_END")]: String(config.rtpEnd),
    [makeToken("CODECS")]: formatCodecs(config.codecs),
    [makeToken("DTMF_MODE")]: config.dtmfMode,
  };
}

function escapeRegExp(literal: string): string {
  return literal.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

const ESC_OPEN_PLACEHOLDER = "__AST_ESC_OPEN__";
const ESC_CLOSE_PLACEHOLDER = "__AST_ESC_CLOSE__";

/** Render a template using a token map. Unknown tokens are preserved; escaped braces are respected. */
export function renderWithTokens(template: string, tokenMap: Record<string, string>): string {
  if (!template) return template;

  // Protect escaped braces before replacement
  // Replace sequences like "\{{" and "\}}" with placeholders so they survive replacements
  let working = template.replace(/\\\{\{/g, ESC_OPEN_PLACEHOLDER).replace(/\\\}\}/g, ESC_CLOSE_PLACEHOLDER);

  const keys = Object.keys(tokenMap);
  if (keys.length > 0) {
    // Sort by length desc to protect against hypothetical overlapping keys
    const sortedKeys = keys.sort((a, b) => b.length - a.length);
    const pattern = new RegExp(sortedKeys.map(escapeRegExp).join("|"), "g");
    working = working.replace(pattern, (match) => tokenMap[match] ?? match);
  }

  // Restore escaped braces
  return working
    .replace(new RegExp(ESC_OPEN_PLACEHOLDER, "g"), "{{")
    .replace(new RegExp(ESC_CLOSE_PLACEHOLDER, "g"), "}}");
}

const TEMPLATE_FILENAMES = ["ari.conf", "pjsip.conf", "extensions.conf", "manager.conf", "queues.conf"] as const;

function getDefaultTemplatesDir(currentDir: string): string {
  // Try build output first, then fall back to source paths during dev
  const candidates = [
    // When running the built main.js: .vite/build -> .vite/infra/asterisk/conf
    path.resolve(currentDir, "../infra/asterisk/conf"),
    // When running in dev (cwd = repo root)
    path.resolve(process.cwd(), "src/main/infra/asterisk/conf"),
    // From compiled dir up to repo root then src
    path.resolve(currentDir, "../../src/main/infra/asterisk/conf"),
  ];
  for (const candidate of candidates) {
    try {
      if (existsSync(candidate)) return candidate;
    } catch {
      // continue
    }
  }
  // Default to the first candidate even if missing; downstream read will throw with context
  return candidates[0];
}

async function readTemplateFile(sourceDir: string, filename: string): Promise<string> {
  const filePath = path.join(sourceDir, filename);
  try {
    return await fs.readFile(filePath, "utf8");
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new Error(`Failed to read template '${filename}' from '${sourceDir}': ${message}`);
  }
}

export async function renderAsteriskConfig(
  config: AsteriskConfig,
  sourceDir?: string,
  targetDir?: string,
  preview = true
): Promise<{ files: Record<string, string>; written: string[] }> {
  const validation = validateAsteriskConfigSchema(config);
  if (!validation.valid) {
    throw new Error(`Invalid Asterisk config: ${validation.errors.join(", ")}`);
  }

  const effectiveSourceDir = sourceDir ?? getDefaultTemplatesDir(__dirname);
  const tokenMap = buildTokenMap(config);

  const files: Record<string, string> = {};
  for (const filename of TEMPLATE_FILENAMES) {
    const template = await readTemplateFile(effectiveSourceDir, filename);
    files[filename] = renderWithTokens(template, tokenMap);
  }

  const written: string[] = [];
  if (!preview && targetDir) {
    await fs.mkdir(targetDir, { recursive: true });
    await Promise.all(
      Object.entries(files).map(async ([filename, content]) => {
        const outPath = path.join(targetDir, filename);
        await fs.writeFile(outPath, content, "utf8");
        written.push(outPath);
      })
    );
  }

  return { files, written };
}

export function validateAsteriskConfig(config: AsteriskConfig): { valid: boolean; errors: string[] } {
  return validateAsteriskConfigSchema(config);
}
