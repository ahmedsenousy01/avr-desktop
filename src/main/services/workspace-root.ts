import fs from "node:fs";
import path from "node:path";
import { app } from "electron";

const DEFAULT_DIR_NAME = "avr-workspace";

let workspaceRootOverride: string | null = null;

function getDefaultWorkspaceRoot(): string {
  const userData = app.getPath("userData");
  return path.join(userData, DEFAULT_DIR_NAME);
}

function ensureDirectoryExists(targetPath: string): void {
  if (!fs.existsSync(targetPath)) {
    fs.mkdirSync(targetPath, { recursive: true });
  }
}

export function getWorkspaceRoot(): string {
  const resolved = workspaceRootOverride ?? getDefaultWorkspaceRoot();
  ensureDirectoryExists(resolved);
  return resolved;
}

// Intended for tests to direct persistence into a temporary directory
export function setWorkspaceRootForTesting(overridePath: string | null): void {
  workspaceRootOverride = overridePath;
}
