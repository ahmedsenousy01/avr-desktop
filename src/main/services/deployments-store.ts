import fs from "node:fs";
import path from "node:path";

import type { PreflightResult } from "@shared/ipc";
import type { AsteriskConfig } from "@shared/types/asterisk";
import type { Deployment } from "@shared/types/deployments";
import type { ASRProviderId, LLMProviderId, STSProviderId, TTSProviderId } from "@shared/types/validation";
import { DeploymentSchema } from "@shared/types/deployments";
import { renderAsteriskConfig } from "@main/services/asterisk-config";
import { getWorkspaceRoot } from "@main/services/workspace-root";

function getDeploymentsRoot(): string {
  const root = getWorkspaceRoot();
  const dir = path.join(root, "deployments");
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function generateId(): string {
  const rand = Math.random().toString(36).slice(2, 8);
  return `dep_${Date.now().toString(36)}_${rand}`;
}

function slugify(input: string): string {
  return (
    input
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/(^-|-$)+/g, "")
      .slice(0, 64) || "deployment"
  );
}

export type CreateDeploymentArgs =
  | {
      name?: string;
      type: "modular";
      providers: { llm: LLMProviderId; asr: ASRProviderId; tts: TTSProviderId };
    }
  | {
      name?: string;
      type: "sts";
      providers: { sts: STSProviderId };
    };

export function createDeployment(args: CreateDeploymentArgs): Deployment {
  const id = generateId();
  const baseName = args.name ?? (args.type === "modular" ? "Modular Deployment" : "STS Deployment");
  const slug = slugify(`${baseName}-${id.slice(4, 10)}`);
  const now = new Date().toISOString();

  const deployment: Deployment = {
    id,
    name: baseName,
    slug,
    type: args.type,
    providers:
      args.type === "modular"
        ? {
            llm: args.providers.llm,
            asr: args.providers.asr,
            tts: args.providers.tts,
            sts: undefined,
          }
        : {
            llm: undefined,
            asr: undefined,
            tts: undefined,
            sts: args.providers.sts,
          },
    createdAt: now,
    updatedAt: now,
  };

  const dir = path.join(getDeploymentsRoot(), slug);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const file = path.join(dir, "deployment.json");
  fs.writeFileSync(file, JSON.stringify(deployment, null, 2), "utf8");

  return deployment;
}

export function listDeployments(): Deployment[] {
  const dir = path.join(getDeploymentsRoot());
  if (!fs.existsSync(dir)) return [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const deployments: Deployment[] = [];
  for (const ent of entries) {
    if (!ent.isDirectory()) continue;
    const file = path.join(dir, ent.name, "deployment.json");
    if (!fs.existsSync(file)) continue;
    try {
      const json = JSON.parse(fs.readFileSync(file, "utf8"));
      const parsed = DeploymentSchema.parse(json);
      deployments.push(parsed);
    } catch {
      // skip invalid
    }
  }
  return deployments;
}

export function findDeploymentDirById(id: string): string | null {
  const dir = path.join(getDeploymentsRoot());
  if (!fs.existsSync(dir)) return null;
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const ent of entries) {
    if (!ent.isDirectory()) continue;
    const file = path.join(dir, ent.name, "deployment.json");
    if (!fs.existsSync(file)) continue;
    try {
      const json = JSON.parse(fs.readFileSync(file, "utf8"));
      const parsed = DeploymentSchema.parse(json);
      if (parsed.id === id) return path.join(dir, ent.name);
    } catch {
      // ignore
    }
  }
  return null;
}

export async function updateDeployment(
  id: string,
  patch: { name?: string; providers?: Partial<Deployment["providers"]>; asterisk?: AsteriskConfig }
): Promise<Deployment> {
  const dir = findDeploymentDirById(id);
  if (!dir) throw new Error("Deployment not found");
  const file = path.join(dir, "deployment.json");
  const current = DeploymentSchema.parse(JSON.parse(fs.readFileSync(file, "utf8")));
  const next: Deployment = {
    ...current,
    name: patch.name ?? current.name,
    providers: { ...current.providers, ...(patch.providers ?? {}) },
    asterisk: patch.asterisk ?? current.asterisk,
    updatedAt: new Date().toISOString(),
  };
  fs.writeFileSync(file, JSON.stringify(next, null, 2), "utf8");

  // Emit asterisk conf files if provided
  if (patch.asterisk) {
    const astDir = path.join(dir, "asterisk");
    if (!fs.existsSync(astDir)) fs.mkdirSync(astDir, { recursive: true });
    // Wait for files to be written so callers can assert existence
    await renderAsteriskConfig(patch.asterisk, undefined, astDir, false);
  }
  return next;
}

export function duplicateDeployment(id: string, name?: string): Deployment {
  const dir = findDeploymentDirById(id);
  if (!dir) throw new Error("Deployment not found");
  const current = DeploymentSchema.parse(JSON.parse(fs.readFileSync(path.join(dir, "deployment.json"), "utf8")));
  if (current.type === "modular") {
    const created = createDeployment({
      type: "modular",
      name: name ?? `${current.name} Copy`,
      providers: {
        llm: current.providers.llm as LLMProviderId,
        asr: current.providers.asr as ASRProviderId,
        tts: current.providers.tts as TTSProviderId,
      },
    });
    // Copy asterisk settings and files if present
    if (current.asterisk) {
      const newDir = findDeploymentDirById(created.id);
      if (newDir) {
        const newFile = path.join(newDir, "deployment.json");
        const newJson = JSON.parse(fs.readFileSync(newFile, "utf8"));
        newJson.asterisk = current.asterisk;
        newJson.updatedAt = new Date().toISOString();
        fs.writeFileSync(newFile, JSON.stringify(newJson, null, 2), "utf8");

        const srcAst = path.join(dir, "asterisk");
        const dstAst = path.join(newDir, "asterisk");
        if (fs.existsSync(srcAst)) {
          if (!fs.existsSync(dstAst)) fs.mkdirSync(dstAst, { recursive: true });
          for (const ent of fs.readdirSync(srcAst, { withFileTypes: true })) {
            if (ent.isFile()) {
              fs.copyFileSync(path.join(srcAst, ent.name), path.join(dstAst, ent.name));
            }
          }
        }
      }
    }
    return created;
  }
  const created = createDeployment({
    type: "sts",
    name: name ?? `${current.name} Copy`,
    providers: { sts: current.providers.sts as STSProviderId },
  });
  return created;
}

export function deleteDeployment(id: string): boolean {
  const dir = findDeploymentDirById(id);
  if (!dir) return false;
  fs.rmSync(dir, { recursive: true, force: true });
  return true;
}

// --- Preflight results persistence ------------------------------------------

function writeJsonAtomic(filePath: string, data: string): void {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const tempPath = path.join(dir, `${path.basename(filePath)}.tmp-${process.pid}-${Date.now()}`);
  fs.writeFileSync(tempPath, data, { encoding: "utf8" });
  fs.renameSync(tempPath, filePath);
}

export function getPreflightFilePathByDeploymentId(id: string): string | null {
  const dir = findDeploymentDirById(id);
  if (!dir) return null;
  return path.join(dir, "preflight.json");
}

export function readPreflightResultByDeploymentId(id: string): PreflightResult | null {
  const file = getPreflightFilePathByDeploymentId(id);
  if (!file || !fs.existsSync(file)) return null;
  try {
    const raw = fs.readFileSync(file, "utf8");
    return JSON.parse(raw) as PreflightResult;
  } catch {
    return null;
  }
}

export function writePreflightResultByDeploymentId(id: string, result: PreflightResult): void {
  const file = getPreflightFilePathByDeploymentId(id);
  if (!file) throw new Error("Deployment not found");
  const json = JSON.stringify(result, null, 2);
  writeJsonAtomic(file, json);
}
