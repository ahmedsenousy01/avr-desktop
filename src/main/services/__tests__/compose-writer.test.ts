import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import type { Providers } from "@shared/types/providers";
import { DEFAULT_ASTERISK_CONFIG } from "@shared/types/asterisk";
import { createDefaultProviders } from "@shared/types/providers";
import {
  buildComposeObject,
  getAsteriskConfMounts,
  getAsteriskPortMappings,
  renderSortedYaml,
  writeComposeFile,
} from "@main/services/compose-writer";
import { createDeployment } from "@main/services/deployments-store";
import { setWorkspaceRootForTesting } from "@main/services/workspace-root";

// ComposeSpec is imported from compose-writer to avoid duplicating types.

function mkTempDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "avr-compose-writer-test-"));
}

function rmDirRecursive(target: string): void {
  if (fs.existsSync(target)) {
    fs.rmSync(target, { recursive: true, force: true });
  }
}

function makeProviders(overrides?: Partial<Providers>): Providers {
  const base = createDefaultProviders();
  base.openai.apiKey = "sk-test-openai";
  base.anthropic.apiKey = "sk-test-anthropic";
  base.gemini.apiKey = "sk-test-gemini";
  base.deepgram.apiKey = "sk-test-deepgram";
  base.elevenlabs.apiKey = "sk-test-elevenlabs";
  return { ...base, ...(overrides ?? {}) };
}

describe("compose-writer", () => {
  let tmpRoot: string;

  beforeEach(() => {
    tmpRoot = mkTempDir();
    setWorkspaceRootForTesting(tmpRoot);
  });

  afterEach(() => {
    setWorkspaceRootForTesting(null);
    rmDirRecursive(tmpRoot);
  });

  it("builds modular compose with correct services, env, mounts, and ports", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "modular",
      name: "Demo",
      providers: { llm: "openai", asr: "deepgram", tts: "elevenlabs" },
    });

    const { spec, yaml } = buildComposeObject(dep, providers, DEFAULT_ASTERISK_CONFIG);
    expect(typeof yaml).toBe("string");

    const services = spec.services;
    const expectedServiceNames = [
      `${dep.slug}-core`,
      `${dep.slug}-asterisk`,
      `${dep.slug}-ami`,
      `${dep.slug}-asr-deepgram`,
      `${dep.slug}-tts-elevenlabs`,
      `${dep.slug}-llm-openai`,
    ];
    for (const name of expectedServiceNames) {
      expect(services[name]).toBeTruthy();
      expect(services[name].container_name).toBe(name);
    }

    // Env propagation
    expect(services[`${dep.slug}-llm-openai`].environment?.OPENAI_API_KEY).toBe("sk-test-openai");
    expect(services[`${dep.slug}-tts-elevenlabs`].environment?.ELEVENLABS_API_KEY).toBe("sk-test-elevenlabs");
    expect(services[`${dep.slug}-asr-deepgram`].environment?.DEEPGRAM_API_KEY).toBe("sk-test-deepgram");

    // Asterisk mounts and ports
    const ast = services[`${dep.slug}-asterisk`];
    const mounts = getAsteriskConfMounts();
    for (const m of mounts) expect(ast.volumes).toContain(m);
    const ports = getAsteriskPortMappings(DEFAULT_ASTERISK_CONFIG);
    for (const p of ports) expect(ast.ports).toContain(p);
  });

  it("builds modular compose with google asr/tts without injecting JSON-only env", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "modular",
      providers: { llm: "anthropic", asr: "google", tts: "google" },
    });

    const { spec } = buildComposeObject(dep, providers, DEFAULT_ASTERISK_CONFIG);
    const services = spec.services;
    expect(services[`${dep.slug}-llm-anthropic`].environment?.ANTHROPIC_API_KEY).toBe("sk-test-anthropic");
    // google asr/tts should not include provider-derived API keys
    const asrGoogleEnv = services[`${dep.slug}-asr-google`].environment ?? {};
    const ttsGoogleEnv = services[`${dep.slug}-tts-google`].environment ?? {};
    expect(asrGoogleEnv).not.toHaveProperty("OPENAI_API_KEY");
    expect(asrGoogleEnv).not.toHaveProperty("ANTHROPIC_API_KEY");
    expect(asrGoogleEnv).not.toHaveProperty("GEMINI_API_KEY");
    expect(asrGoogleEnv).not.toHaveProperty("DEEPGRAM_API_KEY");
    expect(ttsGoogleEnv).not.toHaveProperty("OPENAI_API_KEY");
    expect(ttsGoogleEnv).not.toHaveProperty("ANTHROPIC_API_KEY");
    expect(ttsGoogleEnv).not.toHaveProperty("GEMINI_API_KEY");
    expect(ttsGoogleEnv).not.toHaveProperty("ELEVENLABS_API_KEY");
  });

  it("builds sts compose with openai realtime and env", () => {
    const providers = makeProviders();
    const dep = createDeployment({ type: "sts", providers: { sts: "openai-realtime" } });
    const { spec } = buildComposeObject(dep, providers, undefined);
    const services = spec.services;
    expect(services[`${dep.slug}-sts-openai`]).toBeTruthy();
    expect(services[`${dep.slug}-sts-openai`].environment?.OPENAI_API_KEY).toBe("sk-test-openai");
  });

  it("writes docker-compose.yml idempotently", () => {
    const providers = makeProviders();
    const dep = createDeployment({ type: "modular", providers: { llm: "openai", asr: "deepgram", tts: "elevenlabs" } });
    const { filePath, changed } = writeComposeFile(dep, providers, DEFAULT_ASTERISK_CONFIG);
    expect(filePath.endsWith("docker-compose.yml")).toBe(true);
    expect(changed).toBe(true);
    const first = fs.readFileSync(filePath, "utf8");
    const again = writeComposeFile(dep, providers, DEFAULT_ASTERISK_CONFIG);
    expect(again.changed).toBe(false);
    const second = fs.readFileSync(filePath, "utf8");
    expect(first).toBe(second);
  });

  it("sorts YAML keys deterministically", () => {
    const y = renderSortedYaml({ b: { z: 1, a: 2 }, a: [2, 1] });
    const lines = y.trim().split(/\r?\n/);
    // 'a' should come before 'b'
    expect(lines[0].startsWith("a:"));
  });

  it("produces modular YAML containing expected mounts, ports, and env", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "modular",
      name: "Snapshot Modular",
      providers: { llm: "openai", asr: "deepgram", tts: "elevenlabs" },
    });
    const fixed = { ...dep, slug: "snapshot-modular" };
    const { yaml } = buildComposeObject(fixed, providers, DEFAULT_ASTERISK_CONFIG);
    expect(yaml).toContain("snapshot-modular-asterisk");
    expect(yaml).toContain("./asterisk/conf/pjsip.conf:/etc/asterisk/my_pjsip.conf");
    expect(yaml).toContain("10000-10050:10000-10050/udp");
    expect(yaml).toContain("snapshot-modular-llm-openai");
    expect(yaml).toContain("OPENAI_API_KEY");
    expect(yaml).toContain("snapshot-modular-asr-deepgram");
    expect(yaml).toContain("SPEECH_RECOGNITION_MODEL");
    expect(yaml).toContain('networks:\n  snapshot-modular:\n    driver: "bridge"');
    expect(yaml).toContain("ipam");
  });

  it("produces sts YAML containing expected mounts, ports, and env", () => {
    const providers = makeProviders();
    const dep = createDeployment({ type: "sts", name: "Snapshot STS", providers: { sts: "openai-realtime" } });
    const fixed = { ...dep, slug: "snapshot-sts" };
    const { yaml } = buildComposeObject(fixed, providers, DEFAULT_ASTERISK_CONFIG);
    expect(yaml).toContain("snapshot-sts-asterisk");
    expect(yaml).toContain("./asterisk/conf/manager.conf:/etc/asterisk/my_manager.conf");
    expect(yaml).toContain("10000-10050:10000-10050/udp");
    expect(yaml).toContain("snapshot-sts-sts-openai");
    expect(yaml).toContain("OPENAI_API_KEY");
    expect(yaml).toContain("STS_URL");
    expect(yaml).toContain('networks:\n  snapshot-sts:\n    driver: "bridge"');
    expect(yaml).toContain("ipam");
  });

  it("applies environment overrides to all services", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "sts",
      name: "Override Test",
      providers: { sts: "gemini" },
    });

    // Add environment overrides to the deployment
    const depWithOverrides = {
      ...dep,
      environmentOverrides: {
        GEMINI_MODEL: "gemini-2.5-flash-preview-native-audio-dialog",
        GEMINI_INSTRUCTIONS: "You are a helpful assistant",
        CUSTOM_VAR: "test-value",
        OPENAI_MODEL: "gpt-4o", // This should appear in all services but only be used by relevant ones
      },
    };

    const { spec } = buildComposeObject(depWithOverrides, providers, DEFAULT_ASTERISK_CONFIG);
    const services = spec.services;

    // Verify overrides are applied to all services
    Object.values(services).forEach((service) => {
      expect(service.environment?.GEMINI_MODEL).toBe("gemini-2.5-flash-preview-native-audio-dialog");
      expect(service.environment?.GEMINI_INSTRUCTIONS).toBe("You are a helpful assistant");
      expect(service.environment?.CUSTOM_VAR).toBe("test-value");
      expect(service.environment?.OPENAI_MODEL).toBe("gpt-4o");
    });

    // Verify the sts-gemini service specifically has the overrides
    const stsGemini = services[`${dep.slug}-sts-gemini`];
    expect(stsGemini.environment?.GEMINI_MODEL).toBe("gemini-2.5-flash-preview-native-audio-dialog");
    expect(stsGemini.environment?.GEMINI_INSTRUCTIONS).toBe("You are a helpful assistant");
  });

  it("environment overrides take precedence over default values", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "sts",
      name: "Precedence Test",
      providers: { sts: "gemini" },
    });

    // Override the default GEMINI_MODEL
    const depWithOverrides = {
      ...dep,
      environmentOverrides: {
        GEMINI_MODEL: "custom-gemini-model",
        PORT: "9999", // Override default port
      },
    };

    const { spec } = buildComposeObject(depWithOverrides, providers, DEFAULT_ASTERISK_CONFIG);
    const services = spec.services;

    // Verify overrides take precedence
    const stsGemini = services[`${dep.slug}-sts-gemini`];
    expect(stsGemini.environment?.GEMINI_MODEL).toBe("custom-gemini-model");
    expect(stsGemini.environment?.PORT).toBe("9999");
  });

  it("environment overrides work with modular deployments", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "modular",
      name: "Modular Override Test",
      providers: { llm: "openai", asr: "deepgram", tts: "elevenlabs" },
    });

    const depWithOverrides = {
      ...dep,
      environmentOverrides: {
        OPENAI_MODEL: "gpt-4o",
        DEEPGRAM_MODEL: "nova-2",
        ELEVENLABS_VOICE_ID: "custom-voice",
        CUSTOM_VAR: "modular-test",
      },
    };

    const { spec } = buildComposeObject(depWithOverrides, providers, DEFAULT_ASTERISK_CONFIG);
    const services = spec.services;

    // Verify overrides are applied to all services
    Object.values(services).forEach((service) => {
      expect(service.environment?.OPENAI_MODEL).toBe("gpt-4o");
      expect(service.environment?.DEEPGRAM_MODEL).toBe("nova-2");
      expect(service.environment?.ELEVENLABS_VOICE_ID).toBe("custom-voice");
      expect(service.environment?.CUSTOM_VAR).toBe("modular-test");
    });

    // Verify specific services have the relevant overrides
    const llmOpenai = services[`${dep.slug}-llm-openai`];
    expect(llmOpenai.environment?.OPENAI_MODEL).toBe("gpt-4o");

    const asrDeepgram = services[`${dep.slug}-asr-deepgram`];
    expect(asrDeepgram.environment?.DEEPGRAM_MODEL).toBe("nova-2");

    const ttsElevenlabs = services[`${dep.slug}-tts-elevenlabs`];
    expect(ttsElevenlabs.environment?.ELEVENLABS_VOICE_ID).toBe("custom-voice");
  });

  it("handles empty environment overrides gracefully", () => {
    const providers = makeProviders();
    const dep = createDeployment({
      type: "sts",
      name: "Empty Override Test",
      providers: { sts: "gemini" },
    });

    const depWithEmptyOverrides = {
      ...dep,
      environmentOverrides: {},
    };

    const { spec } = buildComposeObject(depWithEmptyOverrides, providers, DEFAULT_ASTERISK_CONFIG);
    const services = spec.services;

    // Should still have default environment variables
    const stsGemini = services[`${dep.slug}-sts-gemini`];
    expect(stsGemini.environment?.GEMINI_MODEL).toBe("gemini-2.5-flash-preview-native-audio-dialog");
    expect(stsGemini.environment?.PORT).toBe("6037");
  });
});
