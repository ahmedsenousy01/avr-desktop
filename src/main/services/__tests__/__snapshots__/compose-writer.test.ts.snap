// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compose-writer > produces stable YAML snapshot for modular selection 1`] = `
"networks:
  snapshot-modular-mfmka3:
    driver: "bridge"
    name: "snapshot-modular-mfmka3"
services:
  snapshot-modular-mfmka3-ami:
    container_name: "snapshot-modular-mfmka3-ami"
    image: "agentvoiceresponse/avr-ami"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "ami"
          - "snapshot-modular-mfmka3-ami"
    restart: "always"
  snapshot-modular-mfmka3-asr-deepgram:
    container_name: "snapshot-modular-mfmka3-asr-deepgram"
    environment:
      DEEPGRAM_API_KEY: "sk-test-deepgram"
    image: "agentvoiceresponse/avr-asr-deepgram"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "asr"
          - "snapshot-modular-mfmka3-asr-deepgram"
    restart: "always"
  snapshot-modular-mfmka3-asterisk:
    container_name: "snapshot-modular-mfmka3-asterisk"
    image: "agentvoiceresponse/avr-asterisk"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "asterisk"
          - "snapshot-modular-mfmka3-asterisk"
    ports:
      - "5038:5038"
      - "5060:5060"
      - "8088:8088"
      - "10000-20000:10000-20000/udp"
    restart: "always"
    volumes:
      - "./asterisk/manager.conf:/etc/asterisk/my_manager.conf:ro"
      - "./asterisk/pjsip.conf:/etc/asterisk/my_pjsip.conf:ro"
      - "./asterisk/extensions.conf:/etc/asterisk/my_extensions.conf:ro"
      - "./asterisk/queues.conf:/etc/asterisk/my_queues.conf:ro"
      - "./asterisk/ari.conf:/etc/asterisk/my_ari.conf:ro"
  snapshot-modular-mfmka3-core:
    container_name: "snapshot-modular-mfmka3-core"
    image: "agentvoiceresponse/avr-core"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "core"
          - "snapshot-modular-mfmka3-core"
    restart: "always"
  snapshot-modular-mfmka3-llm-openai:
    container_name: "snapshot-modular-mfmka3-llm-openai"
    environment:
      OPENAI_API_KEY: "sk-test-openai"
    image: "agentvoiceresponse/avr-llm-openai"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "llm"
          - "snapshot-modular-mfmka3-llm-openai"
    restart: "always"
  snapshot-modular-mfmka3-tts-elevenlabs:
    container_name: "snapshot-modular-mfmka3-tts-elevenlabs"
    environment:
      ELEVENLABS_API_KEY: "sk-test-elevenlabs"
    networks:
      snapshot-modular-mfmka3:
        aliases:
          - "tts"
          - "snapshot-modular-mfmka3-tts-elevenlabs"
    restart: "always"
version: "3.9"
"
`;

exports[`compose-writer > produces stable YAML snapshot for sts selection 1`] = `
"networks:
  snapshot-sts-mfmka3:
    driver: "bridge"
    name: "snapshot-sts-mfmka3"
services:
  snapshot-sts-mfmka3-ami:
    container_name: "snapshot-sts-mfmka3-ami"
    image: "agentvoiceresponse/avr-ami"
    networks:
      snapshot-sts-mfmka3:
        aliases:
          - "ami"
          - "snapshot-sts-mfmka3-ami"
    restart: "always"
  snapshot-sts-mfmka3-asterisk:
    container_name: "snapshot-sts-mfmka3-asterisk"
    image: "agentvoiceresponse/avr-asterisk"
    networks:
      snapshot-sts-mfmka3:
        aliases:
          - "asterisk"
          - "snapshot-sts-mfmka3-asterisk"
    ports:
      - "5038:5038"
      - "5060:5060"
      - "8088:8088"
      - "10000-20000:10000-20000/udp"
    restart: "always"
    volumes:
      - "./asterisk/manager.conf:/etc/asterisk/my_manager.conf:ro"
      - "./asterisk/pjsip.conf:/etc/asterisk/my_pjsip.conf:ro"
      - "./asterisk/extensions.conf:/etc/asterisk/my_extensions.conf:ro"
      - "./asterisk/queues.conf:/etc/asterisk/my_queues.conf:ro"
      - "./asterisk/ari.conf:/etc/asterisk/my_ari.conf:ro"
  snapshot-sts-mfmka3-core:
    container_name: "snapshot-sts-mfmka3-core"
    image: "agentvoiceresponse/avr-core"
    networks:
      snapshot-sts-mfmka3:
        aliases:
          - "core"
          - "snapshot-sts-mfmka3-core"
    restart: "always"
  snapshot-sts-mfmka3-sts-openai:
    container_name: "snapshot-sts-mfmka3-sts-openai"
    environment:
      OPENAI_API_KEY: "sk-test-openai"
    image: "agentvoiceresponse/avr-sts-openai"
    networks:
      snapshot-sts-mfmka3:
        aliases:
          - "sts"
          - "snapshot-sts-mfmka3-sts-openai"
    restart: "always"
version: "3.9"
"
`;
